src/mesh/interpolation/petsc_hermite_spline_xz.cxx

```diff
@@ -0,0 +1,240 @@
+  h00_z.allocate();
+  h01_z.allocate();
+  h10_z.allocate();
+  h11_z.allocate();
+
+  // The stencil has 16 elements, so maximum number of columns in any
+  // given row (whether on local or remote process) is 16
+  // NOLINTNEXTLINE(misc-include-cleaner)
+  MatMPIAIJSetPreallocation(*weights.get(), 16, nullptr, 16, nullptr);
+}
+
+void PetscXZHermiteSpline::calcWeights(const Field3D& delta_x, const Field3D& delta_z,
+                                       const std::string& region) {
+
+  const int ny = localmesh->LocalNy;
+  const int nz = localmesh->LocalNz;
+  const int xend = (localmesh->xend - localmesh->xstart + 1) * localmesh->getNXPE()
+                   + localmesh->xstart - 1;
+
+  // TODO: work out why using `getRegion(region)` directly causes
+  // issues on more than one core
+  const auto actual_region = getRegion(region);
+  BOUT_FOR(i, actual_region) {
+    const int x = i.x();
+    const int y = i.y();
+    const int z = i.z();
+
+    // The integer part of xt_prime, zt_prime are the indices of the cell
+    // containing the field line end-point
+    int i_corn = static_cast<int>(std::floor(delta_x(x, y, z)));
+    k_corner(x, y, z) = static_cast<int>(std::floor(delta_z(x, y, z)));
+
+    // t_x, t_z are the normalised coordinates \in [0,1) within the cell
+    // calculated by taking the remainder of the floating point index
+    BoutReal t_x = delta_x(x, y, z) - static_cast<BoutReal>(i_corn);
+    const BoutReal t_z = delta_z(x, y, z) - static_cast<BoutReal>(k_corner(x, y, z));
+
+    // NOTE: A (small) hack to avoid one-sided differences
+    if (i_corn >= xend) {
+      i_corn = xend - 1;
+      t_x = 1.0;
+    }
+    if (i_corn < localmesh->xstart) {
+      i_corn = localmesh->xstart;
+      t_x = 0.0;
+    }
+
+    k_corner(x, y, z) = ((k_corner(x, y, z) % nz) + nz) % nz;
+
+    // Check that t_x and t_z are in range
+    if ((t_x < 0.0) || (t_x > 1.0)) {
+      throw BoutException(
+          "t_x={:e} out of range at ({:d},{:d},{:d}) (delta_x={:e}, i_corn={:d})", t_x, x,
+          y, z, delta_x(x, y, z), i_corn);
+    }
+
+    if ((t_z < 0.0) || (t_z > 1.0)) {
+      throw BoutException(
+          "t_z={:e} out of range at ({:d},{:d},{:d}) (delta_z={:e}, k_corner={:d})", t_z,
+          x, y, z, delta_z(x, y, z), k_corner(x, y, z));
+    }
+
+    h00_x[i] = (2. * t_x * t_x * t_x) - (3. * t_x * t_x) + 1.;
+    h00_z[i] = (2. * t_z * t_z * t_z) - (3. * t_z * t_z) + 1.;
+
+    h01_x[i] = (-2. * t_x * t_x * t_x) + (3. * t_x * t_x);
+    h01_z[i] = (-2. * t_z * t_z * t_z) + (3. * t_z * t_z);
+
+    h10_x[i] = t_x * (1. - t_x) * (1. - t_x);
+    h10_z[i] = t_z * (1. - t_z) * (1. - t_z);
+
+    h11_x[i] = (t_x * t_x * t_x) - (t_x * t_x);
+    h11_z[i] = (t_z * t_z * t_z) - (t_z * t_z);
+
+    // Need to convert from global indices to local
+    const auto i_c = Field3D::ind_type(
+        (((localmesh->getLocalXIndex(i_corn) * ny) + (y + y_offset)) * nz
+         + localmesh->getLocalZIndex(k_corner(x, y, z))),
```

Comment:
Note that this seems like it would work with Z-splitting, but it does not!

This is inherently incompatible with z-splitting. It only works if ny and nz are the same as the local ...